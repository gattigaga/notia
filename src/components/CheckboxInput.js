import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import TextareaAutosize from "react-autosize-textarea";

import Icon from "components/Icon";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 8px 0px;
  border-top: ${props => (props.isFocus ? "1px" : "0px")} solid #ddd;
  border-bottom: ${props => (props.isFocus ? "1px" : "0px")} solid #ddd;
`;

const Check = styled.input`
  outline: 0px;
  margin-right: 8px;
  cursor: pointer;
`;

const IconPlus = styled(Icon)`
  margin-right: 8px;
  user-select: none;
  color: #888;
  font-size: 18px;
`;

const IconDelete = styled(Icon)`
  background: #888;
  color: white;
  border-radius: 100%;
  font-size: 16px;
  padding: 2px;
  user-select: none;
  margin-left: 16px;
  cursor: pointer;

  &:hover {
    background: #555;
  }
`;

const Label = styled(TextareaAutosize)`
  width: 100%;
  font-family: Roboto Slab;
  font-size: 14px;
  color: black;
  border: 0px;
  outline: 0px;
  resize: none;

  &::placeholder {
    color: #888;
    user-select: none;
  }
`;

Check.displayName = "Check";
Label.displayName = "Label";
IconDelete.displayName = "IconDelete";

class CheckboxInput extends Component {
  componentWillReceiveProps(nextProps) {
    if (nextProps.isFocus) {
      this.label.textarea.focus();
    }
  }

  render() {
    const {
      label,
      isChecked,
      isNew,
      isFocus,
      onFocus,
      onChangeCheck,
      onChangeLabel,
      onClickDelete,
      onPressEnter
    } = this.props;

    return (
      <Container isFocus={isFocus}>
        {isNew ? (
          <IconPlus name="add" />
        ) : (
          <Check type="checkbox" onChange={onChangeCheck} checked={isChecked} />
        )}
        <Label
          value={label}
          placeholder="List Item"
          rows={1}
          innerRef={ref => {
            this.label = ref;
          }}
          autoFocus={isFocus}
          onChange={onChangeLabel}
          onFocus={onFocus}
          onKeyPress={e => {
            if (e.key === "Enter") {
              e.preventDefault();
              onPressEnter();
            }
          }}
        />
        {!isNew &&
          isFocus && <IconDelete name="close" onClick={onClickDelete} />}
      </Container>
    );
  }
}

CheckboxInput.propTypes = {
  label: PropTypes.string.isRequired,
  isChecked: PropTypes.bool,
  isNew: PropTypes.bool,
  isFocus: PropTypes.bool,
  onFocus: PropTypes.func,
  onChangeCheck: PropTypes.func,
  onChangeLabel: PropTypes.func,
  onClickDelete: PropTypes.func,
  onPressEnter: PropTypes.func
};

export default CheckboxInput;
