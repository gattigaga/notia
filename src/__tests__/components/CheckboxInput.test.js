import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";

import CheckboxInput from "components/CheckboxInput";

let component,
  wrapper,
  onFocus,
  onChangeLabel,
  onClickDelete,
  onChangeCheck,
  onPressEnter;

beforeEach(() => {
  onFocus = jest.fn();
  onChangeLabel = jest.fn();
  onClickDelete = jest.fn();
  onChangeCheck = jest.fn();
  onPressEnter = jest.fn();

  component = (
    <CheckboxInput
      label="Learning React"
      onFocus={onFocus}
      onChangeLabel={onChangeLabel}
      onClickDelete={onClickDelete}
      onChangeCheck={onChangeCheck}
      onPressEnter={onPressEnter}
    />
  );

  wrapper = shallow(component);
});

afterEach(() => {
  onFocus.mockReset();
  onChangeLabel.mockReset();
  onClickDelete.mockReset();
  onChangeCheck.mockReset();
  onPressEnter.mockReset();
});

describe("CheckboxInput", () => {
  it("renders correctly", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders as blank new item", () => {
    wrapper.setProps({ isNew: true });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders as checked", () => {
    wrapper.setProps({ isChecked: true });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders as focused", () => {
    const mountWrapper = mount(component);
    mountWrapper.setProps({ isFocus: true });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders with custom label", () => {
    wrapper.setProps({ label: "Learning Redux" });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("calls 'onChangeLabel'", () => {
    const event = {
      target: {
        value: "Learn SVG"
      }
    };

    expect(onChangeLabel.mock.calls.length).toEqual(0);
    wrapper.find("Label").simulate("change", event);
    expect(onChangeLabel.mock.calls.length).toEqual(1);
    expect(onChangeLabel).toBeCalledWith(event);
  });

  it("calls 'onClickDelete'", () => {
    const mountWrapper = mount(component);
    mountWrapper.setProps({
      isNew: false,
      isFocus: true
    });

    expect(onClickDelete.mock.calls.length).toEqual(0);
    mountWrapper.find("IconDelete").simulate("click");
    expect(onClickDelete.mock.calls.length).toEqual(1);
  });

  it("calls 'onPressEnter'", () => {
    const mountWrapper = mount(component);
    const event = {
      preventDefault() {},
      key: "Enter"
    };

    mountWrapper.setProps({
      isNew: false,
      isFocus: true
    });

    expect(onPressEnter.mock.calls.length).toEqual(0);
    wrapper.find("Label").simulate("keyPress", event);
    expect(onPressEnter.mock.calls.length).toEqual(1);
  });

  it("calls 'onChangeCheck'", () => {
    const mountWrapper = mount(component);
    mountWrapper.setProps({
      isNew: false,
      isFocus: true
    });

    expect(onChangeCheck.mock.calls.length).toEqual(0);
    wrapper.find("Check").simulate("change");
    expect(onChangeCheck.mock.calls.length).toEqual(1);
  });

  it("calls 'onFocus'", () => {
    expect(onFocus.mock.calls.length).toEqual(0);
    wrapper.find("Label").simulate("focus");
    expect(onFocus.mock.calls.length).toEqual(1);
  });
});
