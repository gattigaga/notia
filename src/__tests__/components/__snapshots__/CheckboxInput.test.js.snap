// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxInput renders as blank new item 1`] = `
<styled.div>
  <Styled(Icon)
    name="add"
  />
  <Label
    innerRef={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="List Item"
    rows={1}
    value="Learning React"
  />
</styled.div>
`;

exports[`CheckboxInput renders as checked 1`] = `
<styled.div>
  <Check
    checked={true}
    onChange={[Function]}
    type="checkbox"
  />
  <Label
    innerRef={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="List Item"
    rows={1}
    value="Learning React"
  />
</styled.div>
`;

exports[`CheckboxInput renders as focused 1`] = `
<CheckboxInput
  isFocus={true}
  label="Learning React"
  onChangeCheck={[Function]}
  onChangeLabel={[Function]}
  onClickDelete={[Function]}
  onFocus={[Function]}
  onPressEnter={[Function]}
>
  <styled.div
    isFocus={true}
  >
    <div
      className="sc-bdVaJa bTQhwg"
    >
      <Check
        onChange={[Function]}
        type="checkbox"
      >
        <input
          className="sc-bwzfXH eICtOu"
          onChange={[Function]}
          type="checkbox"
        />
      </Check>
      <Label
        autoFocus={true}
        innerRef={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        placeholder="List Item"
        rows={1}
        value="Learning React"
      >
        <TextareaAutosize
          autoFocus={true}
          className="sc-ifAKCX hFtpBi"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="List Item"
          rows={1}
          value="Learning React"
        >
          <textarea
            autoFocus={true}
            className="sc-ifAKCX hFtpBi"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="List Item"
            rows={1}
            value="Learning React"
          />
        </TextareaAutosize>
      </Label>
      <IconDelete
        name="close"
        onClick={[Function]}
      >
        <Icon
          className="sc-bxivhb ddmtAG"
          name="close"
          onClick={[Function]}
        >
          <i
            className="material-icons sc-bxivhb ddmtAG"
            onClick={[Function]}
          >
            close
          </i>
        </Icon>
      </IconDelete>
    </div>
  </styled.div>
</CheckboxInput>
`;

exports[`CheckboxInput renders as static 1`] = `
<styled.div>
  <Check
    onChange={[Function]}
    type="checkbox"
  />
  <Label
    innerRef={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="List Item"
    readOnly={true}
    rows={1}
    value="Learning React"
  />
</styled.div>
`;

exports[`CheckboxInput renders correctly 1`] = `
<styled.div>
  <Check
    onChange={[Function]}
    type="checkbox"
  />
  <Label
    innerRef={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="List Item"
    rows={1}
    value="Learning React"
  />
</styled.div>
`;

exports[`CheckboxInput renders with custom label 1`] = `
<styled.div>
  <Check
    onChange={[Function]}
    type="checkbox"
  />
  <Label
    innerRef={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="List Item"
    rows={1}
    value="Learning Redux"
  />
</styled.div>
`;
